pkg/shares/share_merging_test.go

```diff
@@ -0,0 +1,146 @@
+			false,
+		},
+		{
+			"two evidence shares",
+			[][]byte{evidenceShareStart, evidenceShareContinuation},
+			[]ShareSequence{{NamespaceID: appconsts.EvidenceNamespaceID, Shares: []Share{evidenceShareStart, evidenceShareContinuation}}},
+			false,
+		},
+		{
+			"one message share",
+			[][]byte{messageOneStart},
+			[]ShareSequence{{NamespaceID: messageOneNamespace, Shares: []Share{messageOneStart}}},
+			false,
+		},
+		{
+			"two message shares",
+			[][]byte{messageOneStart, messageOneContinuation},
+			[]ShareSequence{{NamespaceID: messageOneNamespace, Shares: []Share{messageOneStart, messageOneContinuation}}},
+			false,
+		},
+		{
+			"two messages with two shares each",
+			[][]byte{messageOneStart, messageOneContinuation, messageTwoStart, messageTwoContinuation},
+			[]ShareSequence{
+				{NamespaceID: messageOneNamespace, Shares: []Share{messageOneStart, messageOneContinuation}},
+				{NamespaceID: messageTwoNamespace, Shares: []Share{messageTwoStart, messageTwoContinuation}},
+			},
+			false,
+		},
+		{
+			"one transaction, one evidence, one message",
+			[][]byte{transactionShareStart, evidenceShareStart, messageOneStart},
+			[]ShareSequence{
+				{NamespaceID: appconsts.TxNamespaceID, Shares: []Share{transactionShareStart}},
+				{NamespaceID: appconsts.EvidenceNamespaceID, Shares: []Share{evidenceShareStart}},
+				{NamespaceID: messageOneNamespace, Shares: []Share{messageOneStart}},
+			},
+			false,
+		},
+		{
+			"one transaction, one evidence, two messages",
+			[][]byte{transactionShareStart, evidenceShareStart, messageOneStart, messageTwoStart},
+			[]ShareSequence{
+				{NamespaceID: appconsts.TxNamespaceID, Shares: []Share{transactionShareStart}},
+				{NamespaceID: appconsts.EvidenceNamespaceID, Shares: []Share{evidenceShareStart}},
+				{NamespaceID: messageOneNamespace, Shares: []Share{messageOneStart}},
+				{NamespaceID: messageTwoStart, Shares: []Share{messageTwoStart}},
+			},
+			false,
+		},
+		{
+			"one share with invalid size",
+			[][]byte{invalidShare},
+			[]ShareSequence{},
+			true,
+		},
+		{
+			"message one start followed by message two continuation",
+			[][]byte{messageOneStart, messageTwoContinuation},
+			[]ShareSequence{},
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ParseShares(tt.shares)
+			if tt.expectErr && err == nil {
+				t.Errorf("ParseShares() error %v, expectErr %v", err, tt.expectErr)
+			}
+			if reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ParseShares() got %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func generateRawShare(namespace namespace.ID, isMessageStart bool) (rawShare []byte) {
+	infoByte, _ := NewInfoByte(appconsts.ShareVersion, isMessageStart)
+
+	rawShare = append(rawShare, namespace...)
+	rawShare = append(rawShare, byte(infoByte))
+	rawShare = append(rawShare, bytes.Repeat([]byte{0}, appconsts.ShareSize-len(rawShare))...)
```

Comment:
[optional/related to the above comment]
I know it shouldn't matter, but testing with data other than zero data seems like a good idea

Reply:
Good idea, resolved in https://github.com/celestiaorg/celestia-app/pull/828/commits/2dad6e56ceb12e8de1423aa9cab2ad1fc52b5e1f